components/entities/api/src/main/java/org/phenotips/entities/PrimaryEntity.java

```diff
@@ -53,7 +55,14 @@
      *
      * @return a valid document reference
      */
-    DocumentReference getDocument();
+    DocumentReference getDocumentReference();
+
+    /**
+     * Returns the document where the entity is stored.
+     *
+     * @return a valid document reference
+     */
+    XWikiDocument getDocument();
```

Comment:
That is a good point that the controllers need to be changed anyway. I did consider this and I think you are correct. So even though it in this case it is only a theoretical concern, we should be careful going forward about making major changes to widely-used APIs for all of the reasons we discussed in this thread.

As for maintaining custom versions into the future: we are definitely capable of doing this, and for some clients we already do. However, we want to minimize the amount of work required to make old modules work with new versions of PT, because the work that's required to keep these modules up-to-date presents another strain on our already limited resources (at both Gene42 and CCM).